### PR TITLE
File types IDE changes

### PR DESCRIPTION
**Pull request by RikkiGibson**
Related to #60819

@dibarbet for review. This gets us to the min-bar for IDE support.

The SymbolKey changes are important here. Without them EnC will not work at all with the feature. With them we can do some simple EnC without crashing in the IDE. It will still fail when files are added or removed during an EnC session but we anticipate solving that after merging the feature to main, before RTM. #61999

### CODE REVIEW

**Comment by RikkiGibson** [comment]
Just realized this won't pass CI until #61646 is merged 😅. Should still be reviewable though.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.NamedTypeSymbolKey.cs [hunk 1]
```diff
@@ -68,6 +69,9 @@ public static SymbolKeyResolution Resolve(SymbolKeyReader reader, out string? fa
                 return CreateResolution(result, $"({nameof(NamedTypeSymbolKey)} failed)", out failureReason);
             }
 
+            // <(ContainingType)>F(Ordinal)__(SourceName)
+            private static readonly Regex s_fileTypeNamePattern = new Regex(@"<[a-zA-Z_0-9]*>F\d+__", RegexOptions.Compiled);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.NamedTypeSymbolKey.cs:73]: hrmmmmmm :)

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.NamedTypeSymbolKey.cs [hunk 2]
```diff
@@ -76,12 +80,32 @@ public static SymbolKeyResolution Resolve(SymbolKeyReader reader, out string? fa
                 bool isUnboundGenericType,
                 ITypeSymbol[] typeArguments)
             {
-                foreach (var type in container.GetTypeMembers(GetName(metadataName), arity))
+                var nameWithoutArity = removeArity(metadataName);
+                // Need to do a "decoding" step to get a file type source name from its metadata name
+                var sourceName = s_fileTypeNamePattern.Match(nameWithoutArity) is { Success: true, Length: var length }
+                    ? nameWithoutArity.Substring(length)
+                    : nameWithoutArity;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.NamedTypeSymbolKey.cs:87]: is there no `.IsFile` symbol API yet?  i really don't like detecting based on random encoding details of these names.

Comment (RikkiGibson) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.NamedTypeSymbolKey.cs:87]: I'm not a fan either. We are in a position where we have a metadata name and need to get a source name in order to do a lookup. How can we do that sanely? I think that just knowing that `.IsFile` returned true isn't enough. We still have to figure out what source name to pass to `.GetTypeMembers` to get possible symbols to resolve the SymbolKey.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.NamedTypeSymbolKey.cs:87]: > I'm not a fan either. We are in a position where we have a metadata name and need to get a source name in order to do a lookup. How can we do that sanely?

if we know it's a file-type, we can make a new key which encodes its filepath, containing symbol, name and arity.  Then,  resolution can resolve the containing symbool, and lookup the child type in that namespace that has that name/arity and which is located in the specified file.

File: src/EditorFeatures/CSharpTest/SymbolKey/SymbolKeyTests.cs [hunk 3]
```diff
@@ -19,6 +19,81 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.SymbolId
...
+            var solution = workspace.CurrentSolution;
+            var project = solution.Projects.Single();
+
+            var compilation = await project.GetCompilationAsync();
+            var type = compilation.GetTypeByMetadataName("<C>F0__C1");
+            Assert.NotNull(type);
+            var symbolKey = SymbolKey.Create(type);
+            var resolved = symbolKey.Resolve(compilation).Symbol;
+            Assert.Same(type, resolved);
+        }
```
Comment (Youssef1313) [inline src/EditorFeatures/CSharpTest/SymbolKey/SymbolKeyTests.cs:53]: Consider adding a test for a class nested in the file class.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.NamedTypeSymbolKey.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.NamedTypeSymbolKey.cs]: Have merged in the changes from #62218 and added the suggested public API.

File: src/EditorFeatures/CSharpTest/SymbolKey/SymbolKeyCompilationsTests.cs [hunk 4]
```diff
@@ -239,6 +239,94 @@ public void M<Z>(List<Z> list)
...
+        {
+            var src1 = @"using System;
+
+namespace N1.N2
+{
+    file class C { }
+}
+";
+            // this should result in two entirely separate file symbols.
+            // note that the IDE can only distinguish file type symbols with the same name when they have distinct file paths.
```
Comment (cston) [inline src/EditorFeatures/CSharpTest/SymbolKey/SymbolKeyCompilationsTests.cs:291]: >note that the IDE can only distinguish file type symbols [](http://example.com/codeflow?start=15&length=56)

Are we asserting that we found distinct file type symbols? Should we include `Assert.Equal(3, originalSymbols.Length);`?

**Review by cston — APPROVED** [review]

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs [hunk 5]
```diff
@@ -684,5 +684,130 @@ class C
...
+        }
+
+        [Fact]
+        public async Task DoNotSuggestFileAfterFile()
+        {
+            var markup = $$"""
+                file $$
+                """;
+
+            await VerifyItemIsAbsentAsync(markup, "file");
```
Comment (davidwengier) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs:705]: Should we verify that `class` is present here? Or is that covered by existing tests?

**Review by cston — APPROVED** [review]

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/FileKeywordRecommender.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+    public FileKeywordRecommender()
+        : base(SyntaxKind.FileKeyword)
+    {
+    }
+
+    protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
+    {
+        return context.ContainingTypeDeclaration == null
+            && context.IsTypeDeclarationContext(s_validModifiers, SyntaxKindSet.AllTypeDeclarations, canBePartial: true, cancellationToken);
```
Comment (Youssef1313) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/FileKeywordRecommender.cs:28]: `AllTypeDeclarations` doesn't include delegates. Do we get auto completion for `$$ delegate void M();`

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Utilities/SyntaxKindSet.cs [hunk 7]
```diff
@@ -14,6 +14,7 @@ internal class SyntaxKindSet
         public static readonly ISet<SyntaxKind> AllTypeModifiers = new HashSet<SyntaxKind>(SyntaxFacts.EqualityComparer)
             {
                 SyntaxKind.AbstractKeyword,
+                SyntaxKind.FileKeyword,
```
Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Utilities/SyntaxKindSet.cs:17]: Is this change covered by tests?

File: src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs [hunk 8]
```diff
@@ -84,7 +86,8 @@ IMethodSymbol or
                     isVolatile: field?.IsVolatile == true,
                     isExtern: symbol.IsExtern,
                     isAsync: method?.IsAsync == true,
-                    isRequired: symbol.IsRequired());
+                    isRequired: symbol.IsRequired(),
+                    isFile: (symbol as INamedTypeSymbol)?.IsFile == true);
```
Comment (Youssef1313) [inline src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs:90]: Unrelated to file:

I see this is missing other modifiers. Is it intentional?

File: src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs [hunk 9]
```diff
@@ -203,6 +211,7 @@ private enum Modifiers
             Volatile    = 1 << 14,
             Extern      = 1 << 15,
             Required    = 1 << 16,
+            File        = 1 << 17,
```
Comment (Youssef1313) [inline src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs:214]: Since you added this API, consider fixing up:

Consider also updating

https://github.com/dotnet/roslyn/blob/5510ef93015c7123d20c2e336c94549bd2116a76/src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs#L1491-L1492

https://github.com/dotnet/roslyn/blob/5510ef93015c7123d20c2e336c94549bd2116a76/src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs#L1444-L1466

File: src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs [hunk 10]
```diff
@@ -178,6 +183,9 @@ public DeclarationModifiers WithIsExtern(bool isExtern)
         public DeclarationModifiers WithIsRequired(bool isRequired)
             => new(SetFlag(_modifiers, Modifiers.Required, isRequired));
 
+        public DeclarationModifiers WithIsFile(bool isFile)
+            => new(SetFlag(_modifiers, Modifiers.Required, isFile));
```
Comment (Youssef1313) [inline src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs:187]:
```suggestion
            => new(SetFlag(_modifiers, Modifiers.File, isFile));
```

File: src/Compilers/Core/Portable/Symbols/INamedTypeSymbol.cs [hunk 11]
```diff
@@ -56,6 +56,11 @@ public interface INamedTypeSymbol : ITypeSymbol
         /// </summary>
         bool IsComImport { get; }
 
+        /// <summary>
+        /// Indicates the type is declared in source and is only visible in the file it is declared in.
+        /// </summary>
+        bool IsFile { get; }
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Symbols/INamedTypeSymbol.cs:62]: :memo: For the record, the API review for this is tracked by [https://github.com/dotnet/roslyn/issues/62254](https://github.com/dotnet/roslyn/issues/62254)

Comment (jcouv) [inline src/Compilers/Core/Portable/Symbols/INamedTypeSymbol.cs:62]: For new public APIs, we need to have corresponding compiler tests

**Review by jcouv — COMMENTED** [review]
Done with review pass, compiler changes only (iteration 14)

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs (hunk 5 above)
Comment (RikkiGibson) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs:705]: Remarkably, this is not really the case for lots of different modifiers. Try taking an empty file and type `static cl$$`. You won't get a completion :)

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Utilities/SyntaxKindSet.cs (hunk 7 above)
Comment (RikkiGibson) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Utilities/SyntaxKindSet.cs:17]: Added test 'SuggestFileAfterReadonly'

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/FileKeywordRecommender.cs (hunk 6 above)
Comment (RikkiGibson) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/FileKeywordRecommender.cs:28]: Added test 'SuggestFileBeforeDelegate'

File: src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs (hunk 9 above)
Comment (RikkiGibson) [reply src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs:214]: It actually wasn't clear to me whether 'file' should be added to this. I assume if the IDE has need to generate syntax for a file type, it would use this. But I don't know if the use case exists. It would be helpful to know whether I should delete and just stub out a `bool IsFIle => false;` implementation.

**Review by cston — APPROVED** [review]

File: src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs (hunk 9 above)
Comment (Youssef1313) [reply src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs:214]: @RikkiGibson This is a public API. So it's very necessary that 'file' is added here (not necessarily in this PR)

A common SyntaxGenerator pattern is `generator.WithModifiers(node, generator.GetModifiers(node).WithX(true))`

If the node with a class declaration for `file class C { }`, and we called `WithSealed(true)` (for example), I expect the resulting node will drop `file` if the change is not made.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs:214]: @Youssef1313  is correct.  thanks! :)

**Review by CyrusNajmabadi — APPROVED** [review]
lgtm with the changes suggested by @Youssef1313 .  thanks!

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs:705]: well thats' bad.  we should def fix that :D

File: src/Compilers/Core/Portable/Symbols/INamedTypeSymbol.cs (hunk 11 above)
Comment (RikkiGibson) [reply src/Compilers/Core/Portable/Symbols/INamedTypeSymbol.cs:62]: Done

File: src/EditorFeatures/Test/CodeGeneration/CodeGenerationTests.CSharp.cs [hunk 12]
```diff
@@ -220,6 +220,59 @@ public static class C
...
+                await TestAddNamedTypeAsync(input, expected,
+                    accessibility: accessibility,
+                    modifiers: new Editing.DeclarationModifiers(isFile: true));
+            }
+
+            [Theory, Trait(Traits.Feature, Traits.Features.CodeGeneration), WorkItem(544405, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/544405")]
+            [InlineData("struct", TypeKind.Struct)]
+            [InlineData("interface", TypeKind.Interface)]
+            [InlineData("enum", TypeKind.Enum)]
+            public async Task AddFileType(string kindString, TypeKind typeKind)
```
Comment (RikkiGibson) [inline src/EditorFeatures/Test/CodeGeneration/CodeGenerationTests.CSharp.cs:262]: I wasn't able to figure out how to properly test 'delegate', 'record', or 'record struct' here.

- 'delegate' got me a test crash with a stacktrace consisting only of async thunks..
- I didn't see how to specify 'record' and 'record struct' in here.

**Review by Youssef1313 — COMMENTED** [review]
I expect the lack of file being handled here to cause problems with a public API path:

https://github.com/dotnet/roslyn/blob/1fcfda544da9cb4980da03d46bc753424172b75a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpAccessibilityFacts.cs#L126-L143

**Comment by RikkiGibson** [comment]
Are you able to identify a specific editor feature/scenario which will behave incorrectly? Otherwise I may prefer to file an issue 
and address it after this PR.

**Comment by Youssef1313** [comment]
> Are you able to identify a specific editor feature/scenario which will behave incorrectly? I may prefer to file an issue and address it after this PR.

I'll take a look

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs (hunk 5 above)
Comment (davidwengier) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs:705]: <shocked pikachu face>

**Review by jcouv — APPROVED** [review]
Compiler-side LGTM Thanks (iteration 20)
